### PR TITLE
Make line-length linter less noisy

### PR DESCRIPTION
**Pull request by thumbnail**
## Brief

Adds merge-threshold option which merges line-length errors which are close to each other (probable to be the same code-block).

The old behaviour is still configurable by passing 0 as the threshold

## QA plan

run `project-formatter` before and after this change.

### Before
image

### After
image

## Author checklist

* [x] I have QAed the functionality
* [x] The PR has a reasonably reviewable size and a meaningful commit history
* [x] I have run the [branch formatter](https://github.com/nedap/formatting-stack/blob/332a419034ab46fad526a5592f4257353bd695b6/src/formatting_stack/branch_formatter.clj) and observed no new/significative warnings
* [x] The build passes
* [x] I have self-reviewed the PR prior to assignment
* Additionally, I have code-reviewed iteratively the PR considering the following aspects in isolation:
  * [x] Correctness
  * [ ] Robustness (red paths, failure handling etc)
  * [ ] Test coverage
  * [ ] Spec coverage
  * [ ] Documentation
  * [ ] Security
  * [ ] Performance
  * [ ] Breaking API changes
  * [x] Cross-compatibility (Clojure/ClojureScript)

## Reviewer checklist

* [x] I have checked out this branch and reviewed it locally, running it
* [x] I have QAed the functionality
* [x] I have reviewed the PR
* Additionally, I have code-reviewed iteratively the PR considering the following aspects in isolation:
  * [ ] Correctness
  * [ ] Robustness (red paths, failure handling etc)
  * [ ] Test coverage
  * [ ] Spec coverage
  * [ ] Documentation
  * [ ] Security
  * [ ] Performance
  * [ ] Breaking API changes
  * [ ] Cross-compatibility (Clojure/ClojureScript)

### CODE REVIEW

**Comment by vemv** [comment]
Could you cover that the improvement works?

**Review by tcoenraad — APPROVED** [review]
QA ✅ 

I was wondering if we can do the message still a bit better though.

Suggestion: something like `Line exceeding 130 columns (and 4 lines more)`

**Comment by thumbnail** [comment]
That's a good idea! I wondered about adding row-end col-end to the reporting (clj-kondo reports these too, so we could report a block if possible).


Note how the next 4 lines are not necessarily exceeding the lines, but at least line 1 and line 4 are (line 2 and 3 are collapsed into the block). Because code exceeding the column 'limit' is often grouped together (deeply nested `let`, etc). 


I'll see if i can improve the reporting :)

**Comment by thumbnail** [comment]
@tcoenraad I fixed the reporting, see OP.

A test has been added too.

File: src/formatting_stack/linters/line_length.clj [hunk 1]
```diff
@@ -3,28 +3,50 @@
...
+                     (conj memo report))))
+               [])
+       (mapv (fn [{:keys [line line-start] :as report}]
+               (assoc report
+                      :msg (str "Line exceeding " max-line-length " columns"
+                                (when (not= line line-start)
+                                  (str " (spanning " (inc (- line line-start)) " lines)"))
+                                ".")
+              ;; make sure reporting is on the first line (not the last)
+                      :line line-start)))))
```
[the remark this answers is not in the document]
Comment (thumbnail) [reply src/formatting_stack/linters/line_length.clj]: Thanks @lennartbuit for the `partition-between` suggestion

File: src/formatting_stack/linters/line_length.clj [hunk 2]
```diff
@@ -2,29 +2,43 @@
...
-                           :line     (inc i)
-                           :msg      (str "Line exceeding " threshold " columns.")}))))
-       (filterv some?)))
+(defn exceeding-lines [{:keys [max-line-length merge-threshold]} filename]
+  (->> (-> filename slurp string/split-lines)
+       (map count)
+       (keep-indexed (fn [i column]
+                       (when (< max-line-length column)
+                         {:column column
+                          :line   (inc i)})))
```
Comment (lennartbuit) [inline src/formatting_stack/linters/line_length.clj:15]: Sneaky. But why do we need to increment line by one :').

File: src/formatting_stack/util.clj [hunk 3]
```diff
@@ -159,3 +159,11 @@
...
                *err* s#]
        ~@body)))
+
+;; see https://stackoverflow.com/a/23221442/2046200
+(defn partition-between
+  "Applies f to each value and the next value, splitting each time f returns a new value.
+  Returns a lazy sequence or partitions."
+  [pred coll]
+  (let [switch (reductions not= true (map pred coll (rest coll)))]
+    (map (partial map first) (partition-by second (map list coll switch)))))
```
Comment (lennartbuit) [inline src/formatting_stack/util.clj:169]: 🤯

**Review by lennartbuit — APPROVED** [review]

File: src/formatting_stack/linters/line_length.clj (hunk 2 above)
Comment (lennartbuit) [reply src/formatting_stack/linters/line_length.clj:15]: Ah, okay, we are creating like a report here by looping over the lines, and zero indexed ah yeah.

Comment (thumbnail) [reply src/formatting_stack/linters/line_length.clj:15]: Yeah the collection is zero-indexed; most (all?) editors are one-indexed. so compensating for that here.

**Pull request merged**